Officialize validate refactor (#4114)
demisto/demisto-sdk: demisto_sdk/commands/validate/README.md

**Use Cases**
This command is used to make sure that the content repo files are valid and are able to be processed by Demisto.
This is used in our validation process both locally and in Circle CI.
This command is used to make sure that the content repo files are valid and are able to be processed by the platform.
This is used in our validation process both locally and in gitlab.

**Arguments**:
* **--no-backward-comp**
Whether to check backward compatibility or not.
* **--no-conf-json**
Skip conf.json validation.
* **-s, --id-set**
Perform validations using the id_set file.
* **-idp, --id-set-path**
The path of the id-set.json used for validations.
* **--prev-ver**
Previous branch or SHA1 commit to run checks against.
* **-g, --use-git**
Validate changes using git - this will check the current branch's changes against origin/master.
If the **--post-commit** flag is supplied: validation will run only on the current branch's changed files that have been committed.
If the **--post-commit** flag is not supplied: validation will run on all changed files in the current branch, both committed and not committed.

> * **-pc, --post-commit**

Reviewer: all these flags still work in the old validate flow right? We should for the time being still have docs for them, maybe under a "deprecated" section or in another file that we link to from here

Author: how about renaming the old readme to old_Validate_readme.md with all the info related to the old validate?

Author: Will add a new old_validate_readme.md file and will link it to the readme.md

This will validate all files under the content pack `HelloWorld`.

`demisto-sdk validate --run-old-validate --skip-new-validate -a`
This will validate all files in the repo using the old validate method.

`demisto-sdk validate --config-path {config_file_path} -a`
This will validate all files in the repo using the settings configured in the config file in the given path.

### Error Codes and Ignoring Them
Each error found by validate  has an error code attached to it - the code can be found in brackets preceding the error itself.
For example: `path/to/file: [IN103] - The type field of the proxy parameter should be 8`
In addition, each pack has a `.pack-ignore` file. In order to ignore a certain validation for a given file, the error-code needs to be listed in the **ignorable_errors** section in the config-file (more about this in the config file section), and the user need to mention the file name (only the name and extension, without the whole path), and the error code to ignore.
For example: This .pack-ignore will not fail ipinfo_v2.yml on the validations with the codes BA108 & BA109.
[file:ipinfo_v2.yml]
ignore=BA108,BA109

### Config file

> Each user will have a personal config file which he can edit however he wants.

Reviewer: we need to also give an example config file which is complex, so not the default config

Author: The default config includes all the features but just with a small amount of validations

A default config file can be found [here.](default_config.toml)
The default config file cover all the mandatory validations - the validations that without them the upload will fail.
The config file will have a **ignorable_errors** - A list of the error codes that can be ignored in the .pack-ignore file.
The config file will also have different handling categories:
**validate_all** - the configurations to run when running with -a / -i flags.
**use_git** - the configurations to run when running with -g flag.
And custom sections provided by the user - will be configured to run with the **category-to-run** flag.

> Each handling category will have the following options:

Reviewer: what is a "handling category" ?

> Each user will have a personal config file which he can edit however he wants.
> A default config file can be found [here.](default_config.toml)
> The default config file cover all the mandatory validations - the validations that without them the upload will fail.
> The config file will have a **ignorable_errors** - A list of the error codes that can be ignored in the .pack-ignore file.
> The config file will also have different handling categories:
> **validate_all** - the configurations to run when running with -a / -i flags.
> **use_git** - the configurations to run when running with -g flag.
> And custom sections provided by the user - will be configured to run with the **category-to-run** flag.
> Each handling category will have the following options:
> **select** - The validations to run.
> **warning** - Validations to only throw warning (shouldn't fail the flow).
> The config file can also configure which validations to ignore based on the content item support level using the section header support_level.<support_type> where support_type is one of  xsoar, partner, or community.
> If the user wishes to ignore this feature in some of the calls, he can use the **--ignore-support-level** flag.

Reviewer: ```suggestion
You can define a config file to suit your bussiness needs. If no file is defined, the  [default config file](default_config.toml) will be used.
The default configuration covers basic validations, which prevent successful upload of the content to XSOAR.
#### How to define a configuration file
You can define the following sections:
**ignorable_errors** - a list of the error codes that can be ignored for individual content items in the .pack-ignore file.
**validate_all** - the configurations to run when running with -a / -i flags.
**use_git** - the configurations to run when running with -g flag.
You can also define custom sections - which can be configured to run with the **category-to-run** flag.
Each handling category will have the following options:
**select** - The validations to run.
**warning** - Validations for which to only throw warnings (will not fail the flow).
The config file can also configure which validations to ignore based on the content item support level using the section header support_level.<support_type> where support_type is one of  xsoar, partner, or community.
If the user wishes to ignore this feature in some of the calls, he can use the **--ignore-support-level** flag.
```

This file contains information about our new validate flow, for more information about the old validate flow, please refer to [old_validate_readme](demisto_sdk/commands/validate/old_validate_readme.md)

## Validate

Makes sure your content repository files are in order and have valid file scheme.

> This file 

Reviewer 2: should this be deleted?

Author: Yap, that's fine.

If the **--post-commit** flag is not supplied: validation will run on all changed files in the current branch, both committed and not committed.
* **-a, --validate-all**
Whether to run all validation on all files or not.
* **-i, --input**
The path of the content pack/file to validate specifically.
* **-pc, --post-commit**
Whether the validation should run only on the current branch's committed changed files. This applies only when the **-g** flag is supplied.
* **-st, --staged**
Whether the validation should run only on the current branch's staged files. This applies only when the **-g** flag is supplied.
* **--prev-ver**
Previous branch or SHA1 commit to run checks against.
* **--no-multiprocessing**
Run validate all without multiprocessing, for debugging purposes.
* **-j, --json-file**
The JSON file path to which to output the command results.
* **--category-to-run**
Run specific validations by stating category they're listed under in the config file.
* **-f, --fix**
Wether to autofix failing validations with an available auto fix or not.
* **--config-path**

> Path for a config file to run, if not given - will run the default path at: [demisto_sdk/commands/validate/default_config.toml](default_config.toml)

Reviewer 2: ```suggestion
Path for a config file to run. If not given - will run the default path at: [demisto_sdk/commands/validate/default_config.toml](default_config.toml)
```

* **--ignore-support-level**

> Wether to skip validations based on their support level or not.
> * **--run-old-validate**
> "Wether to run the old validate flow or not. Alteratively, you can configure the RUN_OLD_VALIDATE env variable.
> * **--skip-new-validate**
> Wether to skip the new validate flow or not. Alteratively, you can configure the SKIP_NEW_VALIDATE env variable.

Reviewer 2: ```suggestion
Whether to skip validations based on their support level or not.
* **--run-old-validate**
"Whether to run the old validate flow or not. Alternatively, you can configure the RUN_OLD_VALIDATE env variable.
* **--skip-new-validate**
Whether to skip the new validate flow or not. Alternatively, you can configure the SKIP_NEW_VALIDATE env variable.
```

**Examples**:

`demisto-sdk validate --prev-ver SHA1-HASH`
This will validate only changed files from the branch given (SHA1).

`demisto-sdk validate --post-commit`
This indicates that the command runs post commit.

`demisto-sdk validate -i Packs/HelloWorld/Integrations/HelloWorld/HelloWorld.yml`
This will validate the file Packs/HelloWorld/Integrations/HelloWorld/HelloWorld.yml only.

`demisto-sdk validate -a`

> This will validate all files under `Packs` directory.

Reviewer 2: ```suggestion
This will validate all files under the `Packs` directory.
```

`demisto-sdk validate -i Packs/HelloWorld`
This will validate all files under the content pack `HelloWorld`.

`demisto-sdk validate --run-old-validate --skip-new-validate -a`
This will validate all files in the repo using the old validate method.

`demisto-sdk validate --config-path {config_file_path} -a`
This will validate all files in the repo using the settings configured in the config file in the given path.

### Error Codes and Ignoring Them

> Each error found by validate  has an error code attached to it - the code can be found in brackets preceding the error itself.

Reviewer 2: ```suggestion
Each error found by validate has an error code attached to it. The code can be found in brackets preceding the error itself.
```

For example: `path/to/file: [IN103] - The type field of the proxy parameter should be 8`

> In addition, each pack has a `.pack-ignore` file. In order to ignore a certain validation for a given file, the error-code needs to be listed in the **ignorable_errors** section in the config-file (more about this in the config file section), and the user need to mention the file name (only the name and extension, without the whole path), and the error code to ignore.

Reviewer 2: ```suggestion
In addition, each pack has a `.pack-ignore` file. In order to ignore a certain validation for a given file, the error-code needs to be listed in the **ignorable_errors** section in the config-file (see the [Config file section](#config-file)), and the user needs to mention the file name (only the name and extension, without the whole path), and the error code to ignore.
```

For example: This .pack-ignore will not fail ipinfo_v2.yml on the validations with the codes BA108 & BA109.
[file:ipinfo_v2.yml]
ignore=BA108,BA109

### Config file

> You can define a config file to suit your bussiness needs. If no file is defined, the  [default config file](default_config.toml) will be used.

Reviewer 2: ```suggestion
You can define a config file to suit your business needs. If no file is defined, the  [default config file](default_config.toml) will be used.
```

> The default configuration covers basic validations, which prevent successful upload of the content to XSOAR.

Reviewer 2: ```suggestion
The default configuration covers basic validations, which prevents successful upload of the content to Cortex XSOAR.
```
Why would you want to prevent successful uploads? If the content passes all validations, don't you want it to upload?
Maybe it should say
The default configuration covers basic validations, which prevents successful upload of the **unvalidated** content to Cortex XSOAR.

Author: Acctually it prevents failures in the upload.
So maybe:
'The default configuration covers basic validations, which prevents unsuccessful uploads of the validated content to Cortex XSOAR.'
wdyt? @ShirleyDenkberg

Reviewer 2: sounds good. 
The default configuration covers basic validations, which prevents unsuccessful uploads of the validated content to Cortex XSOAR.
